### PR TITLE
Persist ExposureWindow based risk level calculation results and update the UI (EXPOSUREAPP-3910,EXPOSUREAPP-3855)

### PR DESCRIPTION
**Pull request by d4rken**
Combined efforts of @chris-cwa and me to rid ourselves of the `RisklevelRepository` and make UI updates more consistent by providing a single-source-of-truth for risk level related data.

* Remove `RiskLevelRespository`
* Replace static access to `LocalData.*risklevel*`
* Refactor all access to the results produced by the  `RiskLevelTask`. The task stores the data in `RiskLevelStorage`, and any component wants the calculated score can get the latest scores via `RiskLevelStorage.latestResults: Flow<List<RiskLevelResult>>`.
    * UI Components subscribe to changes in `RiskLevelStorage`
    * Notifications are displayed via `RiskLevelChangeDetector` by monitoring new `RiskLevelResult`s

### TODO:
* [x] Tests
* [x] Switch to new UI texts added bei #1697 
* [x] Legacy data migration
* [x] Persist ExposureWindow's in tester builds.

### Testing
* Use the debug menu to fake a high/low/default risklevel state and  check that the riskcard looks like the screenshots in EXPOSUREAPP-3855
* Test general risk level behavior on fresh installs and when time traveling

New risk card examples:

Screenshot from 2020-11-24 12-58-40
Screenshot from 2020-11-24 12-58-23
Screenshot from 2020-11-24 12-58-10

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator.kt
Comment (chris-cwa) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator.kt:30]: if (time == 0) return null

**Review by janetback — DISMISSED** [review]

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
Comment (kolyaopahle) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt:139]: This also feels like something that could be part of riskLevelStorage?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt
Comment (kolyaopahle) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt:29]: Why is this commented out and not removed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt
Comment (kolyaopahle) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:32]: Is this actually monitoring the app config?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelDataTest.kt
Comment (kolyaopahle) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelDataTest.kt:25]: since we are actually testing RiskLevelSettings maybe rename/refactor the test?

File: Corona-Warn-App/src/main/res/values-bg/strings.xml
Comment (harambasicluka) [inline Corona-Warn-App/src/main/res/values-bg/strings.xml:37]: shouldn't be excluded to not cause any merge conflicts as the translation delivery will "override" everything. In the newest translation delivery after this is merged this shouldn't be included. But as we'll get our next translation delivery tomorrow I would say we leave it as it is.

Comment (harambasicluka) [reply Corona-Warn-App/src/main/res/values-bg/strings.xml:37]: We may have to do resolve some merge conflicts on the translation delivery.

Comment (harambasicluka) [reply Corona-Warn-App/src/main/res/values-bg/strings.xml:37]: This applies for all languages != de

**Review by harambasicluka — COMMENTED** [review]
Added some questions, have to review the tests.

Of course naming is highly optional, but I would prefer if we could stick to `HIGH` or `INCREASED` Risk in our code.

**Review by harambasicluka — CHANGES_REQUESTED** [review]
Updated 1.7.0/RC8 to this branch, App crashed instantly.

```
E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-4 @coroutine#69
    Process: de.rki.coronawarnapp.test, PID: 31644
    java.lang.NoClassDefFoundError: Failed resolution of: Lde/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator$getLegacyResults$1$1;
        at de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator.getLegacyResults(RiskLevelResultMigrator.kt:48)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$3$2.emit(Collect.kt:136)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$2$2.emit(Collect.kt:136)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(Unknown Source:4)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:82)
        (Coroutine boundary)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1.invokeSuspend(CoroutinesRoom.kt:82)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1.invokeSuspend(CoroutinesRoom.kt:75)
        at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:212)
        at kotlinx.coroutines.flow.internal.CombineKt$asFairChannel$1.invokeSuspend(Combine.kt:143)
        at kotlinx.coroutines.flow.FlowKt__EmittersKt$onStart$$inlined$unsafeFlow$1.collect(Emitters.kt:120)
        at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1.collect(Emitters.kt:114)
        at androidx.lifecycle.FlowLiveDataConversions$asLiveData$1.invokeSuspend(FlowLiveData.kt:139)
        at androidx.lifecycle.BlockRunner$maybeRun$1.invokeSuspend(CoroutineLiveData.kt:176)
     Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lde/rki/coronawarnapp/risk/storage/legacy/RiskLevelResultMigrator$getLegacyResults$1$1;
        at de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator.getLegacyResults(RiskLevelResultMigrator.kt:48)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$3$2.emit(Collect.kt:136)
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$2$2.emit(Collect.kt:136)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(Unknown Source:4)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55)
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:82)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator$getLegacyResults$1$1" on path: DexPathList[[zip file "/data/app/~~fvwqdjrF-VLWwJljB4CPig==/de.rki.coronawarnapp.test-_ENM36BVz0C2KLhOLp3G2w==/base.apk"],nativeLibraryDirectories=[/data/app/~~fvwqdjrF-VLWwJljB4CPig==/de.rki.coronawarnapp.test-_ENM36BVz0C2KLhOLp3G2w==/lib/arm64, /data/app/~~fvwqdjrF-VLWwJljB4CPig==/de.rki.coronawarnapp.test-_ENM36BVz0C2KLhOLp3G2w==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64, /product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:207)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at de.rki.coronawarnapp.risk.storage.legacy.RiskLevelResultMigrator.getLegacyResults(RiskLevelResultMigrator.kt:48) 
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$3$2.emit(Collect.kt:136) 
        at de.rki.coronawarnapp.risk.storage.BaseRiskLevelStorage$$special$$inlined$map$2$2.emit(Collect.kt:136) 
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15) 
        at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(Unknown Source:4) 
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:73) 
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:55) 
        at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:82) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665) 
```

**Comment by harambasicluka** [comment]
App is now opening, but the the contact row is missing:
image

**Comment by harambasicluka** [comment]
> App is now opening, but the the contact row is missing:

Still get the same state after a clean install with time traveling to get a low risk.

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
Comment (d4rken) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt:139]: Hm, it's not risk level related though. Downloading and calculation is independent.
In any case, the actual timestamps are already tracked by `KeyPackageSyncSettings`, we could possibly remove the `LocalData.lastTimeDiagnosisKeys` soon, let's evaluate it after all PRs are merged for 1.8.x, what do you think?

**Comment by harambasicluka** [comment]
> > App is now opening, but the the contact row is missing:
> 
> Still get the same state after a clean install with time traveling to get a low risk.

Is "fixed" english strings are empty... you have to use german as your device language :D

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/ExposureStateUpdateWorker.kt:29]: Oversight. Fixed in next commit :+1:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:32]: Copy paste ;)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:59]: Was previously like that, didn't touch that part.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt:90]: Yes

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:74]: It's a scope label.

https://kotlinlang.org/docs/reference/this-expressions.html

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelTask.kt:77]: Will sync with @SamuraiKek and solve the conflicts, let's merge this first.

**Comment by harambasicluka** [comment]
Discussed with @d4rken my questions and bugs are resolved. As I won't be available this evening you can see this PR as approved from my side when Circle Ci passes.

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt
Comment (kolyaopahle) [reply Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/risklevel/ui/TestRiskLevelCalculationFragmentCWAViewModel.kt:139]: hmm true, maybe riskLevelStorage is the wrong place, reevaluation after the 1.8 merge sounds good to me :)

**Review by kolyaopahle — APPROVED** [review]
LGTM 👍

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelResultTest.kt
Comment (BMItter) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/risk/RiskLevelResultTest.kt:22]: little typo here- ristLevel 
but no urgent

**Comment by BMItter** [comment]
Did a deeplook yesterday evening. Code lgtm & could not find any bugs :ok_hand: . left one little comment (optional).

**Review by BMItr — APPROVED** [review]

**Review by harambasicluka — APPROVED** [review]

**Pull request merged**
